internal/agent/agent.go

```diff
@@ -46,64 +55,116 @@ func (a *Agent) Run(ctx context.Context) error {
 
 	defer utilruntime.HandleCrash()
-	ctx, cancel := context.WithCancel(ctx)
-	shutdownSignals := []os.Signal{os.Interrupt, syscall.SIGTERM}
-	// handle teardown
-	shutdownHandler := make(chan os.Signal, 2)
-	signal.Notify(shutdownHandler, shutdownSignals...)
-	// health check closing ch
-	healthCheckCh := make(chan chan any)
-	go func(ctx context.Context) {
-		select {
-		case <-shutdownHandler:
-			a.log.Infof("Received SIGTERM or SIGINT signal, shutting down.")
-			//We must wait for the health checker to close any open requests and the log file.
-			c := make(chan any)
-			healthCheckCh <- c
-			<-c
-			a.log.Infof("Health check stopped.")
-
-			close(shutdownHandler)
-			cancel()
-		case <-ctx.Done():
-			a.log.Infof("Context has been cancelled, shutting down.")
-			//We must wait for the health checker to close any open requests and the log file.
-			c := make(chan any)
-			healthCheckCh <- c
-			<-c
-			a.log.Infof("Health check stopped.")
-
-			close(shutdownHandler)
-			cancel()
-		}
-	}(ctx)
-
-	StartServer(a.log, a.config)
 
 	client, err := newPlannerClient(a.config)
 	if err != nil {
 		return err
 	}
 
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, syscall.SIGHUP, syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT)
+
+	ctx, cancel := context.WithCancel(ctx)
+	a.start(ctx, client)
+
+	<-sig
+
+	a.log.Info("stopping agent...")
+
+	a.Stop()
+	cancel()
+
+	return nil
+}
+
+func (a *Agent) Stop() {
+	serverCh := make(chan any)
+	a.server.Stop(serverCh)
+
+	<-serverCh
+	a.log.Info("server stopped")
+
+	c := make(chan any)
+	a.healtCheckStopCh <- c
+	<-c
+	a.log.Info("health check stopped")
+}
+
+func (a *Agent) start(ctx context.Context, plannerClient client.Planner) {
+	// start server
+	a.server = NewServer(defaultAgentPort, a.config.DataDir, a.config.WwwDir)
+	go a.server.Start(a.log)
+
+	// get the credentials url
+	a.initializeCredentialUrl()
+
 	// start the health check
 	healthChecker, err := NewHealthChecker(
 		a.log,
-		client,
+		plannerClient,
 		a.config.DataDir,
 		time.Duration(a.config.HealthCheckInterval*int64(time.Second)),
 	)
 	if err != nil {
-		return err
+		a.log.Fatalf("failed to start health check: %w", err)
 	}
-	healthChecker.Start(healthCheckCh)
+
+	// TODO refactor health checker to call it from the main goroutine
+	healthChecker.Start(a.healtCheckStopCh)
 
 	collector := NewCollector(a.log, a.config.DataDir)
 	collector.collect(ctx)
 
-	inventoryUpdater := NewInventoryUpdater(a.log, a.config, client)
-	inventoryUpdater.UpdateServiceWithInventory(ctx)
+	inventoryUpdater := NewInventoryUpdater(a.log, a.config, a.credUrl, plannerClient)
+	updateTicker := jitterbug.New(time.Duration(a.config.UpdateInterval.Duration), &jitterbug.Norm{Stdev: 30 * time.Millisecond, Mean: 0})
 
-	return nil
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-updateTicker.C:
+			}
+
+			//	check for health. Send requests only if we have connectivity
```

Comment:
Can you please explain why the code below is not part of `case <-updateTicker.C:`?

Reply:
It is but I've put it out of the scope for better readability. the `select` will block until tick or cancel().